jaeger-core/src/test/java/com/uber/jaeger/samplers/PerOperationSamplerTest.java

```diff
@@ -0,0 +1,156 @@
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import com.uber.jaeger.samplers.http.OperationSamplingParameters;
+import com.uber.jaeger.samplers.http.PerOperationSamplingParameters;
+import com.uber.jaeger.samplers.http.ProbabilisticSamplingStrategy;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PerOperationSamplerTest {
+
+  private final double SAMPLING_RATE = 0.31415;
+  private final double DEFAULT_SAMPLING_PROBABILITY = 0.512;
+  private final double DEFAULT_LOWER_BOUND_TRACES_PER_SECOND = 2.0;
+  private final int MAX_OPERATIONS = 100;
+  private final long traceId = 1L;
+  private final String operation = "some operation";
+
+  @Mock private ProbabilisticSampler defaultProbabilisticSampler;
+  private HashMap<String, GuaranteedThroughputSampler>  operationToSamplers = new HashMap<>();
+  private PerOperationSampler undertest;
+
+  @Before
+  public void setUp(){
+    undertest = new PerOperationSampler(MAX_OPERATIONS, operationToSamplers, defaultProbabilisticSampler);
+    when(defaultProbabilisticSampler.sample(operation, traceId)).thenReturn(SamplingStatus.of(true, new HashMap<String, Object>()));
+  }
+
+  @After
+  public void tearDown() {
+    undertest.close();
+  }
+
+  @Test
+  public void testFallbackToDefaultProbabilisticSampler() {
+    undertest = new PerOperationSampler(0, operationToSamplers, defaultProbabilisticSampler);
+    SamplingStatus samplingStatus = undertest.sample(operation, traceId);
+    assertTrue(samplingStatus.isSampled());
+    verify(defaultProbabilisticSampler).sample(operation, traceId);
+  }
+
+  @Test
+  public void testCreateGuaranteedSamplerOnSample() {
+    String newOperation = "new operation";
+    undertest.sample(newOperation, traceId);
+    assertEquals(new GuaranteedThroughputSampler(0, 0), operationToSamplers.get(newOperation));
+  }
+
+  @Test
+  public void testPerOperationSampler() {
+    GuaranteedThroughputSampler guaranteedThroughputSampler = mock(GuaranteedThroughputSampler.class);
+    operationToSamplers.put(operation, guaranteedThroughputSampler);
+
+    when(guaranteedThroughputSampler.sample(operation, traceId)).thenReturn(SamplingStatus.of(true, new HashMap<String, Object>()));
+
+    SamplingStatus samplingStatus = undertest.sample(operation, traceId);
+    assertTrue(samplingStatus.isSampled());
+    verify(guaranteedThroughputSampler).sample(operation, traceId);
+    verifyNoMoreInteractions(defaultProbabilisticSampler);
```

Comment:
+1, but conversely a test is missing when you do fall back onto default sampler (when maxOps limit is reached)